### ISSUE TITLE
Does setting a COMPlus_ environment variable after the application has already started take effect?

### ISSUE THREAD

**Issue by avparuch** [dotnet/runtime#57713]
I'm trying to set a COMPlus environment variable. https://github.com/steveharter/dotnet_coreclr/blob/master/Documentation/project-docs/clr-configuration-knobs.md

This is my code:

```
using System;
using System.Threading;

namespace minthreads
{
    class Program
    {
        static void Main(string[] args)
        {
            Environment.SetEnvironmentVariable("COMPlus_ThreadPool_ForceMinWorkerThreads", "0x32");
            ThreadPool.GetMinThreads(out int workerThreads, out int ioThreads);
            Console.WriteLine(workerThreads);
            Console.WriteLine(ioThreads);
        }
    }
}
```

The workerThreads variable does not return 50. It returns 8, which is the number of processors on my machine. Does this mean coreCLR does not read the environment variables after the application has already started.

However, setting the same environment variable via command line or powershell like this $env:COMPlus_ThreadPool_ForceMinWorkerThreads = '0x32' before the application is started does work: the workerThreads does return 50.

**Comment by jkotas** [comment]
> Does this mean coreCLR does not read the environment variables after the application has already started.

Correct. All environment variables that affect the runtime has to be set before the process starts.

**Comment by jkotas** [comment]
You can use `ThreadPool.SetMinThreads` to change the min threads setting in running process.

**Comment by avparuch** [comment]
> > Does this mean coreCLR does not read the environment variables after the application has already started.
> 
> Correct. All environment variables that affect the runtime has to be set before the process starts.

Thank you for the confirmation.

**Comment by avparuch** [comment]
> You can use `ThreadPool.SetMinThreads` to change the min threads setting in running process.

When I use the `ThreadPool.SetMinThreads` API, the thread pool still starts off at processor count. It just grows unthrottled till the minthreads count and then settles down to its normal algorithm if/when it crosses the minthreads threshold. I'm trying to force a minimum from the get-go.. Are there other techniques to warm up the thread pool to a certain number of threads before application starts taking a burst of traffic?

**Comment by avparuch** [comment]
> > Are there other techniques to warm up the thread pool to a certain number of threads
> 
> This will force the threadpool to create certain number of threads:
> 
> ```cs
> WarmUpThreadPool(100);
> Console.WriteLine(ThreadPool.ThreadCount);
> 
> static void WarmUpThreadPool(int threads)
> {
>     ThreadPool.SetMinThreads(threads, threads);
>     var tasks = new Task[threads];
>     var barrier = new Barrier(threads);
>     for (int i = 0; i < tasks.Length; i++) tasks[i] = Task.Run(() => barrier.SignalAndWait());
>     Task.WaitAll(tasks);
> }
> ```

Thank you, works like a charm!
Also, I noticed that after X seconds, the threadpool count settles down to an appropriate steady state level. It scales down from the warmed up threadpool count, if there is no work to do. Is there a way to force a minimum number of threads to always be there even if there is no work to do? (in anticipation of future work).

**Comment by jkotas** [comment]
>  Is there a way to force a minimum number of threads to always be there even if there is no work to do?

I do not know. @kouvel ?

**Comment by mangod9** [comment]
@avparuch are you noticing that for your scenario the thread "creation" is causing measurable perf degradation after a period of no activity?

**Comment by kouvel** [comment]
> Is there a way to force a minimum number of threads to always be there even if there is no work to do? (in anticipation of future work).

There isn't a way to force the threads to remain running at the moment. Threads exit after about 20 seconds of being inactive currently.

**Comment by avparuch** [comment]
> > Is there a way to force a minimum number of threads to always be there even if there is no work to do? (in anticipation of future work).
> 
> There isn't a way to force the threads to remain running at the moment. Threads exit after about 20 seconds of being inactive currently.

Yes, I'm noticing that around 20th second of inactivity, threads exit. Thank you for confirming.

**Comment by kouvel** [comment]
It would probably make sense to keep at least one thread around, though I'm not sure how much it would help. On my Windows machine it takes around 200 us to create a thread and around 5 us to release a waiting thread. It's a large difference but still may be unnoticeable for latency. Did you find a case where it would help to keep threads around?
